File: haskell/lib/Free/Core/Interpreters/Stepper.hs
module Free.Core.Interpreters.Stepper (appToStepperInterpreter) where

import Free.Core (AppDSL (..), CoinFlipDSL (FlipCoin), ConsoleDSL (PrintLine), runWith, type (~>))
import Free.Stepper.Core (Stepper, askBoundedEnum, notifyStepAndWait)

-- One can get creative with the possible interpreters, this one for example translates
-- the App DSL to the Stepper one so that it could be interpreted with the ad hoc stepper
-- interpreters

appToStepperInterpreter :: AppDSL ~> Stepper
appToStepperInterpreter = \case
  EvalCoinFlip c k -> k <$> c `runWith` coinToStepperInterpreter
  EvalConsole c k -> k <$> c `runWith` consoleToStepperInterpreter

coinToStepperInterpreter :: CoinFlipDSL ~> Stepper
coinToStepperInterpreter = \case
  FlipCoin k -> k <$> askBoundedEnum "What is the flip result? "

consoleToStepperInterpreter :: ConsoleDSL ~> Stepper
consoleToStepperInterpreter = \case
  PrintLine msg k -> k <$> notifyStepAndWait msg

File: scala/src/main/scala/free/core/interpreters/Stepper.scala
package free.core.interpreters

import cats.~>
import free.core.*
import free.core.AppDSL.*
import free.core.CoinFlipDSL.*
import free.core.ConsoleDSL.*
import free.stepper.Core.*
import free.stepper.Core.Stepper.*
import cats.kernel.BoundedEnumerable.catsKernelBoundedEnumerableForBoolean

object Stepper:
  val appToStepperInterpreter: AppDSL ~> Stepper = new (AppDSL ~> Stepper):
    override def apply[A](a: AppDSL[A]): Stepper[A] = a match
      case EvalCoinFlip(c) => c.foldMap(stepperCoinFlipInterpreter)
      case EvalConsole(c)  => c.foldMap(stepperConsoleInterpreter)

  val stepperCoinFlipInterpreter: CoinFlipDSL ~> Stepper = new (CoinFlipDSL ~> Stepper):
    override def apply[A](c: CoinFlipDSL[A]): Stepper[A] = c match
      case FlipCoin => askBoundedEnum("What is the flip result?")

  val stepperConsoleInterpreter: ConsoleDSL ~> Stepper = new (ConsoleDSL ~> Stepper):
    override def apply[A](c: ConsoleDSL[A]): Stepper[A] = c match
      case PrintLine(msg) => notifyStepAndWait(msg)
